src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,666 @@
+					phase = Phase.OnFinalize;
+				} else {
+					// If this is not a documented phase, it likely has to do with execution initialization
+					// or finalization.
+					phase = // Use camelCase for consistency
+						(secondarySpans[0] && stringCamelCase(secondarySpans[0]?.name)) ||
+						stringCamelCase(primary.name);
+				}
+			}
+
+			const events =
+				eventsByParentId
+					.get(primary.id)
+					?.concat(secondarySpansEvents)
+					.sort((a, b) => a.eventIndex - b.eventIndex) || [];
+
+			const eventsWithPhaseInfo = events.map((e) => {
+				return {
+					...e,
+					phase: {
+						variant: phase,
+						applyExtrinsicIndex: extrinsicIndex,
+					},
+					primarySpanId: {
+						id: primary.id,
+						name: primary.name,
+						target: primary.target,
+					},
+				};
+			});
+
+			operations.push(
+				...this.deriveOperations(
+					this.accountEventsByAddress(eventsWithPhaseInfo)
+				)
+			);
+
+			// Primary span, and all associated descendant spans and events
+			// representing an action group.
+			actions.push({
+				extrinsicIndex,
+				phase,
+				primarySpan: primary,
+				secondarySpans,
+				events: eventsWithPhaseInfo,
+			});
+		}
+
+		return { actions, operations };
+	}
+
+	/**
+	 * Extract an extrinsic index from an `:extrinsic_index` event. If it is not an
+	 * `:extrinsic_index` event returns `null`.
+	 *
+	 * @param event a parsed event
+	 */
+	private maybeExtractIndex(event: PEvent): IOption<BN> {
+		if (
+			!(
+				(event.storagePath as SpecialKeyInfo)?.special === ':extrinsic_index' &&
+				event.data.stringValues.method == 'Get'
+			)
+		) {
+			// Note: there are many read and writes of `:extrinsic_index` per extrinsic
+			// so take care when modifying this logic.
+			// Note: we only consider `Get` ops; `Put` ops are prep for next extrinsic.
+			return null;
+		}
+
+		const indexEncodedOption = event.data.stringValues.result;
+		if (!(typeof indexEncodedOption == 'string')) {
+			throw new InternalServerError(
+				'Expected there to be an encoded extrinsic index for extrinsic index event'
+			);
+		}
+
+		const scale = indexEncodedOption
+			.slice(2) // Slice of leading `Option` bits beacuse they cause trouble
+			.match(/.{1,2}/g) // Reverse the order of the bytes for correct endian-ness
+			?.reverse()
+			.join('');
+		const hex = `0x${scale || ''}`;
+
+		return this.registry.createType('u32', hex).toBn();
+	}
+
+	/**
+	 * Parse events by
+	 * 1) Adding parent span info, event index and the storage item name to each event.
+	 * Also adds the extrinsic index as an integer if there is an `:extrinsic_index` event.
+	 * 2) Create a map of span id => array of children events
+	 * 3) Create a map of span id => to extrinisic index
+	 *
+	 * @param spansById map of span id => to span with its children for all spans
+	 * from tracing the block.
+	 */
+	private parseEvents(
+		spansById: Map<number, SpanWithChildren>
+	): {
+		eventsByParentId: Map<number, PEvent[]>;
```

Comment:
Could this be something you would want to make into an Interface type?

Reply:
You mean type alias? Potentially .. not sure how to decide though.